### ISSUE TITLE
nvim_buf_get_lines should return [] for unloaded buffer

### ISSUE THREAD

**Issue by blueyed** [neovim/neovim#7688]
Given the following t.vim file:

```vim
" nvim_buf_get_lines behaves different to getbufline for unlisted buffer

" Create an unlisted buffer.
let &errorformat = '%f:%m'
laddexpr tempname().':msg'
let llist = getloclist(0)
" echom string(llist)
let unlisted_bufnr = llist[0].bufnr

echom !buflisted(unlisted_bufnr)

echom string(nvim_buf_get_lines(unlisted_bufnr, 0, 1, 0))
echom string(getbufline(unlisted_bufnr, 1, '$'))

exe 'b '.unlisted_bufnr
echom string(nvim_buf_get_lines(unlisted_bufnr, 0, 1, 0))
echom string(getbufline(unlisted_bufnr, 1, '$'))
```

`nvim -u t.vim` will print:
```
1
['']
[]"/tmp/nvimf8F7O2/1" [New File]
['']
['']
```

I think this should behave the same, regarding of if the buffer has been loaded
or not?!

- `nvim --version`: v0.2.3-70-g27f9b1c7b

**Comment by justinmk** [comment]
Is this the same as https://github.com/neovim/neovim/issues/5534 ? No one has sent a patch.

**Comment by blueyed** [comment]
Sounds similar - but it seems like hidden/unlisted is not involved there, is it?

`:help getbufline` explicitly says btw:

> This function works only for loaded buffers.  For unloaded and non-existing buffers, an empty |List| is returned.

**Comment by justinmk** [comment]
In general the API should behave consistently with ... the API There's no stated requirement or hint that it should match VimL functions. Is there some reason the getbufline() behavior is preferable? Note that getbufline() did not exist when nvim_buf_get_lines() was first designed.

**Comment by justinmk** [comment]
The comments in the sample code, compared to the sample output, are confusing.. Can you just say exactly what you expect to happen?

**Comment by wsdjeg** [comment]
@blueyed This script does not work in vim8, I think maybe you should create a script for vim8, BTW, I think neovim's `nvim_buf_get_lines` seems better.

**Comment by wsdjeg** [comment]
Ohh, sorry, I got it, there is some difference between them, @justinmk I think what blueyed want is  `nvim_buf_get_lines` should return an empty list for unloaded and non-existing buffers.

**Comment by wsdjeg** [comment]
@blueyed 

```vim
" nvim_buf_get_lines behaves different to getbufline for unlisted buffer

" Create an unlisted buffer.
let &errorformat = '%f:%m'
laddexpr tempname().':msg'
let llist = getloclist(0)
" echom string(llist)
let unlisted_bufnr = llist[0].bufnr

echom !bufexists(unlisted_bufnr)
```

hmm, but in both vim and neovim, bufexists return 1, so that means the buffer is loaded?

**Comment by blueyed** [comment]
@justinmk 
Fixed/removed the confusing comments.

What I've expected is that `nvim_buf_get_lines` can be used for better performance to get lines of a buffer, and in particular `nvim_buf_line_count` to get the count (this returns 1 also for unloaded buffers).

Re loaded/listed: I think the difference here is about if the buffer is _loaded_ or not.. / can be checked with `bufloaded()`.

**Comment by phodge** [comment]
Perhaps a better solution would be for `nvim_buf_get_lines()` to just temporarily load the buffer if it's not loaded. That way it's always guaranteed to give the correct results. Any thoughts?

**Comment by justinmk** [comment]
Difficult to say. It would load the buffer then immediately unload it? Could be a hidden performance trap. To avoid the trap , clients would check if a buffer is loaded before working with it.

But we could achieve the same pattern by instead throwing an error for unloaded buffers. 

So the question is what is the use case for using nvim_buf_get_lines on an unloaded buffer. @blueyed ?

Seems like the best approach is as requested, return empty list to indicate unloaded. And might be good to add a new function like `nvim_buf_is_loaded`.

**Comment by blueyed** [comment]
> use case for using nvim_buf_get_lines on an unloaded buffer

Nothing in particular IIRC - I've just noted that there is a difference.

> Seems like the best approach is as requested, return empty list to indicate unloaded. And might be good to add a new function like nvim_buf_is_loaded.

:+1:

**Comment by phodge** [comment]
I don't know that there's a use-case for calling `nvim_buf_get_lines()` on an unloaded buffer, it's more that plugins can't be certain whether or not the buffer is loaded and they need an obvious, safe way to handle the scenario. I mean, I think an `nvim_buf_is_loaded()` helper is a good idea and it'd be trivial to implement, but it only pretends to give you safety:

    if nvim_buf_is_loaded(bufnr):
        # race condition! buffer may have been unloaded just now by the user
        lines = nvim_buf_get_lines(bufnr, ...)

Good programmers will notice that there's a race condition there and won't be impressed by our shonky API, and good programmers are the ones we really want to encourage to keep coming back and writing plugins. :-)

**Comment by phodge** [comment]
I do think we should at least provide an option that allows `nvim_buf_get_lines()` to load/unload the buffer automatically for you. I think there _is_ a valid use case for this - e.g. you want to automatically refactor a bunch of files using vim buffers so that undo/redo information is remembered by neovim. But I don't know of any simple way to make sure the buffer is loaded other than opening it in a new window or tab.

I think plugin authors would appreciate being able to just write a one-liner like the following and be done:

    # always works - even if the buffer is unloaded
    lines = `nvim_buf_get_lines(bufnr, ..., autoload=true)`

**Comment by justinmk** [comment]
`nvim_call_atomic` as well as `nvim_execute_lua` are the general (and encouraged) way to avoid such race conditions. (We should of course continue to work on making those approaches more _ergonomic_, but the general problem of race conditions between sequential API calls is solved. Though, it would also be welcome to discuss new approaches to achieving "atomicity" of RPC calls.)

> nvim_buf_get_lines(bufnr, ..., autoload=true)`

We can't change the function signature. See `:help api-contract`.

> plugin authors would appreciate being able to just write a one-liner

Yes, currently our answer to that is for client libs to provide such abstractions. (And the Nvim built-in Lua library could offer similar wrappers.)

> But I don't know of any simple way to make sure the buffer is loaded other than opening it in a new window or tab.

I think @bfredl has mentioned the idea of a `nvim_buf_open` function or something like that.

**Comment by bfredl** [comment]
`nvim_buf_open` was unrelated, but one could imagine`nvim_buf_ensure_loaded`. The trouble is that loading a buffer according to the rules is irreducibly nontrivial: swap files might need to be taken care of, autocmds are expected be fired (at least `BufReadCmd` if the file is not a normal file). A plugin in a situation where it just wants to read a buffer might not want to cause side effects.

Personally I simply never use unloaded buffers, it makes the vim experience simpler. If I reached this situation in one of my own plugins, I would simply have the plugin refuse to operate on an unloaded buffer (unless the operation naturally involves opening the buffer in a split, or something like that)

**Comment by phodge** [comment]
@justinmk: thanks for the extra info, that really helps narrow down the solution space.

So if I want to continue in the vein of "keep the API simple; put sugar in the client libs", then it seems like it'd be most reasonable for the `nvim_buf_get_lines()` to just return a sentinel value or raise an error for unloaded buffers, and then in the python (or whichever) client lib, wrap the call in a helper that looks for the sentinel and turns it into a known exception for you. Then in your python plugin you can handle it in a pythonic way: pretend the error doesn't exist it if you know it won't happen, or wrap it in try/catch so you can handle it gracefully.

So the question then becomes - is the `[]` sentinel value sufficient, or do we need to return an error? I'm still concerned that when you ask for the line range (5, -1) to get everything from line 5 to the end of the file, you don't know whether a return value of `[]` means there's no lines past line 5, or that the buffer was suddenly unloaded.

**Comment by justinmk** [comment]
Yes, `[]` sentinel value sufficient. 

> I'm still concerned that when you ask for the line range (5, -1) to get everything from line 5 to the end of the file, you don't know whether a return value of [] means there's no lines past line 5,

Yeah, but there's nothing we can do. Throwing an error might be something we would choose to do with a new function ,but I think it's too risky without enough benefit to change this existing function.

> nvim_buf_get_lines() - return a proper sentinel value for unloaded buffers (0 or -1?)

should return 0.

**Comment by bfredl** [comment]
> line range (5, -1) to get everything from line 5 to the end of the file,

This or most things like that should throw an error already if strict indexing is active (which new plugins/clients should do) and the buffer is unloaded (the only difference would be `(0, 1)`, which now is allowed I presume, but should throw error with unloaded buffer)

**Comment by phodge** [comment]
@bfredl: looking at the implementation of `normalize_index()`, it doesn't check to see if the buffer is loaded or not. If it's not loaded, it'll think the line count is `1`, so no error gets raised if you're just asking for the first line's contents. This won't be a problem for `nvim_buf_get_lines()` once I change it to bail out early for unloaded buffers, but I probably need to at least investigate what other features are using `normalize_index()` that might be incorrectly assuming the buffer is loaded.

**Comment by justinmk** [comment]
closed in #8660
